backend/api/oauthEnabledClient.ts

```diff
@@ -0,0 +1,108 @@
+import superagent from 'superagent'
+import { Readable } from 'stream'
+import Agent, { HttpsAgent } from 'agentkeepalive'
+import logger from '../../logger'
+import { getHeaders } from './axios-config-decorators'
+import { Context } from '../interfaces/context'
+
+export interface ClientFactoryParams {
+  baseUrl: string
+  timeout: number
+}
+
+export interface OAuthEnabledClient {
+  get: (context: Context, path: string, resultLimit?: number) => Promise<superagent.Response>
+  post: (context: Context, path: string, body?: unknown) => Promise<superagent.Response>
+  put: (context: Context, path: string, body: unknown) => Promise<superagent.Response>
+  del: (context: Context, path: string, body?: unknown) => Promise<superagent.Response>
+}
+
+const resultLogger = (result: superagent.Response) => {
+  logger.debug(`${result.req.method} ${result.req.path} ${result.status}`)
+  return result
+}
+
+const errorLogger = (error: superagent.ResponseError) => {
+  const status = error.response ? error.response.status : '-'
+  const responseData = error.response ? error.response.body : '-'
+  if (error.response && error.response.req) {
+    logger.warn(
+      `API error in ${error.response.req.method} ${error.response.req.path} ${status} ${error.message} ${responseData}`,
+    )
+  } else logger.warn(`API error with message ${error.message}`)
+  return error
+}
+
+export const oauthEnabledClientFactory = (params: ClientFactoryParams): OAuthEnabledClient => {
```

Comment:
meh.. could potentially return this as a class and get rid of the extra interface

(or could derive the interface type using typeof and [Return](https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypetype) )

Superagent returns promises so don't need to wrap,  e.g: [here](https://github.com/ministryofjustice/hmpps-template-typescript/blob/main/server/data/restClient.ts)